src/InfiniteIterableValue.php

```diff
@@ -0,0 +1,311 @@
+     * @return IterableValue
+     */
+    public function flatMap(callable $transformer)
+    {
+        $clone = clone $this;
+        $clone->stack = $clone->stack->push(function (iterable $iterable) use ($transformer) {
+            foreach ($iterable as $value) {
+                yield from $transformer($value);
+            }
+        });
+
+        return $clone;
+    }
+
+    public function toArrayValue(): ArrayValue
+    {
+        return Wrap::array($this->toArray());
+    }
+
+    /**
+     * @param mixed $value
+     * @return IterableValue
+     */
+    public function unshift($value)
+    {
+        $clone = clone $this;
+        $clone->stack = $clone->stack->push(function (iterable $iterable) use ($value) {
+            yield $value;
+            yield from $iterable;
+        });
+
+        return $clone;
+    }
+
+    /**
+     * @param mixed $value
+     * @return IterableValue
+     */
+    public function push($value)
+    {
+        $clone = clone $this;
+        $clone->stack = $clone->stack->push(function (iterable $iterable) use ($value) {
+            yield from $iterable;
+            yield $value;
+        });
+
+        return $clone;
+    }
+
+    /**
+     * @return IterableValue
+     */
+    public function join(IterableValue $other)
+    {
+        $clone = clone $this;
+        $clone->stack = $clone->stack->push(function (iterable $iterable) use ($other) {
+            yield from $iterable;
+            yield from $other;
+        });
+
+        return $clone;
+    }
+
+    /**
+     * @return IterableValue
+     */
+    public function slice(int $offset, int $length)
+    {
+        $clone = clone $this;
+        $clone->stack = $clone->stack->push(function (iterable $iterable) use ($offset, $length) {
+            foreach ($iterable as $value) {
+                if ($offset-- > 0) {
+                    continue;
+                }
+
+                if ($length-- <= 0) {
+                    break;
+                }
+
+                yield $value;
+            }
+        });
+
+        return $clone;
+    }
+
+    /**
+     * @param callable|null $comparator function(mixed $valueA, mixed $valueB): int{-1, 0, 1}
+     * @return IterableValue
+     */
+    public function diff(ArrayValue $other, ?callable $comparator = null)
+    {
+        $clone = clone $this;
+        $clone->stack = $clone->stack->push(function (iterable $iterable) use ($other, $comparator) {
+            $otherValues = $other->toArray();
+
+            foreach ($iterable as $value) {
+                if ($comparator === null) {
+                    $found = \in_array($value, $otherValues, true);
+                } else {
+                    $found = false;
+                    foreach ($otherValues as $otherValue) {
+                        if ($comparator($otherValue, $value) === 0) {
+                            $found = true;
+                            break;
+                        }
+                    }
+                }
```

Comment:
it can be simplified now:

```php
if ($comparator === null) {
    $found = $other->hasElement($value);
} else {
    $found = $other->any(function($otherValue) use ($comparator): bool {
        return $comparator($otherValue, $value) === 0;
    });
}
```